### DISCUSSION TITLE
issue: watched data triggered after with untransformed input

### DISCUSSION THREAD

**Discussion by davidrhoderick** [react-hook-form/react-hook-form#8255] [Q&A] [answered]
### Version Number

7.29.0

### Codesandbox/Expo snack

https://codesandbox.io/s/sad-beaver-2vw7yn

### Steps to reproduce

1. Install dependencies for MUI (including emotion) & React Hook Form.
2. Follow documentation on [transforming & parsing](https://react-hook-form.com/advanced-usage#TransformandParse).
3. Implement with additional props for transformation (transformation is dynamic).
4. Notice that transformation doesn't work internally; value is _visually_ set correctly but the actual `value` is not sliced correctly.  This can be seen with the `watchAllFields` call and when the submit button is clicked.

### Expected behaviour

When following the documentation and implementing the same functionality on both the transform/parse input and output functions, I'd expect that the CVC field is truncated in the `watchAllFields` call _as well as_ the submission.  What drives me crazy is that the watched data is output correctly _first_ only to be output a second time incorrectly.  Is there some event bubbling going on that I need to address?

### What browsers are you seeing the problem on?

Firefox 99.0, Chrome 99.0.4844.84

### Relevant log output

Console logs:

```shell
raw value 1234 // The input value
{code: {size: 3}} // The props used to slice the input
transformed value 123 // The correctly transformed value
watched data card_cvv 123 // The correctly output watched value
watched data card_cvv 1234 // ??? Why is the raw value output a second time ???
{ card_cvv: "1234" } // The handleSubmit logger also outputs the raw value
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment, marked as answer]
Could you try following this codesandbox https://codesandbox.io/s/transform-vt3tm?

I would look into your onChange, see if you have parsed input value, and pass it back to hook form.

```tsx
const { field: { onChange } } = useController()

onChange() // this onChnage will send a value back to hook form, which can debug on your example above. 
```

in addition, you can always try to get your controlled component working correctly first with transform and praise, then simply hook it up with `useController`

https://react-hook-form.com/api/usecontroller

example: https://codesandbox.io/s/usecontroller-own-state-wncet2?file=/src/App.tsx

**Reply by davidrhoderick** [reply]
Thanks a lot @bluebill1049, you pointed me to the answer.  I updated my [Codesandbox](https://codesandbox.io/s/sad-beaver-2vw7yn).  Basically, I wasn't updating the correct `onChange` event.  It needed to be the `Controller`'s one as you pointed out.  So that resulted in my applying the `transform` code in the right place and now it works!

**Reply by bluebill1049** [reply]
That's awesome to hear!
